src/libAtomVM/interop.c

```diff
@@ -226,60 +250,28 @@ InteropFunctionResult interop_iolist_size(term t, size_t *size)
 
     temp_stack_destroy(&temp_stack);
 
-    *size = acc;
     return InteropOk;
 }
 
-InteropFunctionResult interop_write_iolist(term t, char *p)
+static InteropFunctionResult write_string_fold_fun(term t, void *accum)
 {
-    if (term_is_binary(t)) {
+    char **p = (char **) accum;
+    if (term_is_integer(t)) {
+        **p = term_to_int(t);
+        (*p)++;
+    } else if (term_is_binary(t)) {
         int len = term_binary_size(t);
-        memcpy(p, term_binary_data(t), len);
-        return InteropOk;
-    }
-
-    struct TempStack temp_stack;
-    if (UNLIKELY(temp_stack_init(&temp_stack) != TempStackOk)) {
-        return InteropMemoryAllocFail;
-    }
-
-    if (UNLIKELY(temp_stack_push(&temp_stack, t) != TempStackOk)) {
-        temp_stack_destroy(&temp_stack);
-        return InteropMemoryAllocFail;
+        memcpy(*p, term_binary_data(t), len);
+        *p += len;
     }
-
-    while (!temp_stack_is_empty(&temp_stack)) {
-        if (term_is_integer(t)) {
-            *p = term_to_int(t);
-            p++;
-            t = temp_stack_pop(&temp_stack);
-
-        } else if (term_is_nil(t)) {
-            t = temp_stack_pop(&temp_stack);
-
-        } else if (term_is_nonempty_list(t)) {
-            if (UNLIKELY(temp_stack_push(&temp_stack, term_get_list_tail(t)) != TempStackOk)) {
-                temp_stack_destroy(&temp_stack);
-                return InteropMemoryAllocFail;
-            }
-            t = term_get_list_head(t);
-
-        } else if (term_is_binary(t)) {
-            int len = term_binary_size(t);
-            memcpy(p, term_binary_data(t), len);
-            p += len;
-            t = temp_stack_pop(&temp_stack);
-
-        } else {
-            temp_stack_destroy(&temp_stack);
-            return InteropBadArg;
-        }
-    }
-
-    temp_stack_destroy(&temp_stack);
     return InteropOk;
 }
 
+InteropFunctionResult interop_write_iolist(term t, char *p)
```

Comment:
Likewise, this function should be renamed `interop_write_iodata`.

Reply:
Same as above comment.